crates/client/rpc/src/versions/user/v0_8_0/methods/read/get_storage_proof.rs

```diff
@@ -0,0 +1,202 @@
+    let mut arr = [0u8; 32];
+    let slice = &mut BitSlice::from_slice_mut(&mut arr)[5..];
+    slice[..path.len()].copy_from_bitslice(path);
+    Felt::from_bytes_be(&arr)
+}
+
+/// Returns (root hash, nodes)
+fn make_trie_proof<H: StarkHash + Send + Sync>(
+    block_n: u64,
+    trie: &mut GlobalTrie<H>,
+    trie_name: StorageProofTrie,
+    identifier: &[u8],
+    keys: Vec<Felt>,
+) -> RpcResult<(Felt, Vec<NodeHashToNodeMappingItem>)> {
+    let mut keys: Vec<_> = keys.into_iter().map(|f| BitArray::new(f.to_bytes_be())).collect();
+    keys.sort();
+
+    tracing::debug!("Getting trie proof for {trie_name:?} on block {block_n} for n={} keys", keys.len());
+
+    let mut storage = trie
+        .get_transactional_state(BasicId::new(block_n), trie.get_config())
+        .map_err(|err| anyhow::anyhow!("{err:#}"))
+        .or_internal_server_error("Getting transactional state")?
+        .ok_or(StarknetRpcApiError::CannotMakeProofOnOldBlock)?;
+
+    let root_hash = storage
+        .root_hash(identifier)
+        .map_err(|err| anyhow::anyhow!("{err:#}"))
+        .or_internal_server_error("Getting root hash of trie")?;
+
+    let proof = storage
+        .get_multi_proof(identifier, keys.iter().map(|k| &k.as_bitslice()[5..]))
+        .map_err(|err| anyhow::anyhow!("{err:#}"))
+        .or_internal_server_error("Error while making storage multiproof")?;
+
+    // convert the bonsai-trie type to the rpc DTO
+    let converted_proof = proof
+        .0
+        .into_iter()
+        .map(|(node_hash, n)| {
+            let node = match n {
+                mc_db::ProofNode::Binary { left, right } => MerkleNode::Binary { left, right },
+                mc_db::ProofNode::Edge { child, path } => {
+                    MerkleNode::Edge { child, path: path_to_felt(&path), length: path.len() }
+                }
+            };
+            NodeHashToNodeMappingItem { node_hash, node }
+        })
+        .collect();
+
+    Ok((root_hash, converted_proof))
+}
+
+pub fn get_storage_proof(
+    starknet: &Starknet,
+    block_id: BlockId,
+    class_hashes: Option<Vec<Felt>>,
+    contract_addresses: Option<Vec<Felt>>,
+    contracts_storage_keys: Option<Vec<ContractStorageKeysItem>>,
+) -> RpcResult<GetStorageProofResult> {
+    // Pending block does not have a state root, so always fallbacck to latest.
+    let block_id = match block_id {
+        BlockId::Tag(BlockTag::Pending) => BlockId::Tag(BlockTag::Latest),
+        block_id => block_id,
+    };
+
+    let block_n = starknet
+        .backend
+        .get_block_n(&block_id)
+        .or_internal_server_error("Resolving block number")?
+        .ok_or(StarknetRpcApiError::NoBlocks)?;
+
+    let Some(latest) = starknet.backend.get_latest_block_n().or_internal_server_error("Getting latest block in db")?
+    else {
+        return Err(StarknetRpcApiError::BlockNotFound.into());
+    };
+
+    if latest.saturating_sub(block_n) > starknet.storage_proof_config.max_distance {
```

Comment:
This means that in the Default for `storage_proof_config` (`max_distance = 0`) the only possible block on which we can get storage proofs is the current one, right?

Which is the maximum `max_distance` possible? (would be cool if we can explain it relation with snapshots and trielogs - if any)

Reply:
max distance can be usize::MAX if you really wanted to. But we run into the risk of this kind of scenario:

current block is 1000
we have 25 snapshots, one every 4 blocks, meaning the oldest snapshot we have is on block 900
user asks for a storage proof on a block very far into the past, way farther than the oldest snapshot we have, like block_n=500

in this scenario, the endpoint will actually still work - it will in fact answer with a storage proof for block_n=500. But to do so it would have to get the snapshot at block_n=900, and apply the trie log in reverse up until block_n=500.
This max_distance is useful to reject requests like these to avoid getting DoSed in public rpc configurations for example.
For SNOS you could totally put usize::MAX, as with max_tries and max_keys

will add a comment about that

Reply:
where do you think i should put this explanation? There's an explanation in RpcParams::rpc_storage_proof_max_distance that looks like it's enough, what do you think?